server/server.go

```diff
@@ -872,6 +857,177 @@ func (s *server) SendOutboundMessage(ctx context.Context, req *rpcpb.SendOutboun
+	if deadline, ok := ctx.Deadline(); !ok || time.Until(deadline) < DefaultStartTimeout {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(context.Background(), DefaultStartTimeout)
+		_ = cancel // don't call since "start" is async, "curl" may not specify timeout
+		zap.L().Info("received start request with default timeout", zap.String("timeout", DefaultStartTimeout.String()))
+	} else {
+		zap.L().Info("received start request with existing timeout", zap.String("deadline", deadline.String()))
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// If [clusterInfo] is already populated, the server has already been started.
+	if s.clusterInfo != nil {
+		return nil, ErrAlreadyBootstrapped
+	}
+
+	var (
+		pid = int32(os.Getpid())
+		err error
+	)
+
+	rootDataDir, err := os.MkdirTemp(os.TempDir(), "network-runner-root-data")
+	if err != nil {
+		return nil, err
+	}
+
+	s.clusterInfo = &rpcpb.ClusterInfo{
+		Pid:         pid,
+		RootDataDir: rootDataDir,
+		Healthy:     false,
+	}
+
+	zap.L().Info("starting",
+		zap.Int32("pid", pid),
+		zap.String("rootDataDir", rootDataDir),
+	)
+
+	if s.network != nil {
+		return nil, ErrAlreadyBootstrapped
+	}
+
+	s.network, err = newLocalNetwork(localNetworkOptions{
+		rootDataDir: rootDataDir,
+
+		// to block racey restart
+		// "s.network.start" runs asynchronously
+		// so it would not deadlock with the acquired lock
+		// in this "Start" method
+		restartMu: s.mu,
+
+		snapshotsDir: s.cfg.SnapshotsDir,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// blocking load snapshot to soon get not found snapshot errors
+	if err := s.network.loadSnapshot(ctx, req.SnapshotName); err != nil {
+		zap.L().Warn("snapshot load failed to complete", zap.Error(err))
+		s.network = nil
+		s.clusterInfo = nil
+		return nil, err
+	}
+
+	// start non-blocking wait to load snapshot results
+	// the user is expected to poll cluster status
+	go s.network.loadSnapshotWait(ctx)
+
+	// update cluster info non-blocking
+	// the user is expected to poll this latest information
+	// to decide cluster/subnet readiness
+	go func() {
+		zap.L().Info("waiting for local cluster readiness")
+		select {
+		case <-s.closed:
+			return
+		case <-s.network.stopCh:
+			// TODO: fix race from shutdown
+			return
+		case serr := <-s.network.startErrCh:
+			zap.L().Warn("snapshot load failed to complete", zap.Error(serr))
+			s.mu.Lock()
+			s.network = nil
+			s.clusterInfo = nil
+			s.mu.Unlock()
+			return
+		case <-s.network.localClusterReadyCh:
+			s.mu.Lock()
+			s.clusterInfo.NodeNames = s.network.nodeNames
+			s.clusterInfo.NodeInfos = s.network.nodeInfos
+			s.clusterInfo.Healthy = true
+			s.mu.Unlock()
+		}
+		select {
+		case <-s.closed:
+			return
+		case <-s.network.stopCh:
+			return
+		case serr := <-s.network.startErrCh:
+			zap.L().Warn("snapshot load failed to complete", zap.Error(serr))
+			s.mu.Lock()
+			s.network = nil
+			s.clusterInfo = nil
+			s.mu.Unlock()
+			return
+		case <-s.network.customVMsReadyCh:
```

Comment:
What if there are no custom VMs?

Reply:
Hmm. 
1. Shouldn't we check for any existing VMs first at all? Like https://github.com/ava-labs/avalanche-network-runner/blob/58c0e5f1a1ceb3fcf48ced18867330ec8a468f7c/server/server.go#L383, why have to wait for Custom VMs if there are none?
2. Doesn't closing of `customVMsReadyCh` prevent any future custom VMs to be deployed properly?
3. Two subsequent `select` looks odd to me, can we not avoid it ?

Reply:
Done 3. For 1, LoadSnapshot() is not currently receiving requests to start custom VMs. It is just loading a snapshot. There is not currently on server a helper function to deploy a VM on a already started network. Probably next thing to add to the server.